[deploy] Improve error messaging for create_movable

tugsbayasgalan opened the pull request:
Summary: This diff makes sure to give clear error message when user tries to create obj from obj that lives in different session

Test Plan: buck test //caffe2/torch/csrc/deploy:test_deploy

Differential Revision: D31323045

suo commented on torch/csrc/deploy/interpreter/interpreter_impl.h:
```diff
@@ -80,6 +80,9 @@ struct Obj {
   Obj callKwargs(std::unordered_map<std::string, c10::IValue> kwargs);
   bool hasattr(const char* attr);
   Obj attr(const char* attr);
+  InterpreterSessionImpl* getInterpreterSessionImpl() {
```
We should not expose this object's `InterpreterSessionImpl` as a public part of the interface. My suggestion: since `InterpreterSessionImpl` is already a friend class, write an `InterpreterSessionImpl::isOwner(Obj)` and use that instead.

tugsbayasgalan replied:
sounds good

suo commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -160,6 +160,14 @@ ReplicatedObj InterpreterSession::createMovable(Obj obj) {
       "Can only create a movable object when the session was created from an interpreter that is part of a InterpreterManager");
+  auto objIntpSession = obj.getInterpreterSessionImpl();
+
+  if (objIntpSession != impl_.get()) {
```
Probably can just `TORCH_CHECK(objIntpSession == impl_.get())` directly.

tugsbayasgalan replied:
yep. will do

suo commented on torch/csrc/deploy/deploy.cpp:
```diff
@@ -160,6 +160,14 @@ ReplicatedObj InterpreterSession::createMovable(Obj obj) {
+  if (objIntpSession != impl_.get()) {
+    TORCH_CHECK(
+        false,
+        "Cannot create movable from an object that lives in different session");
```
Can we improve the error message to specify which interpreter the object is on, and which interpreter is trying to create the movable object? I think interpreters have ids associated with them, right? That should make debugging easier.

tugsbayasgalan replied:
is it notifyIdx_ in InterpreterSession? Obj has InterpreterSessionImpl as its' field but I don't think it knows anything about its' ID. So I can write an error message with Obj's ID and InterpreterSession's notifyIdx_. But getting the obj's interpreterSession's ID seems bit tricky.

tugsbayasgalan replied:
@suo what do you think about this?

suo reviewed (approved):
okay looks good! Thanks for this :)

tugsbayasgalan committed:
[deploy] Improve error messaging for create_movable (#65955)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/65955

This diff makes sure to give clear error message when user tries to create obj from obj that lives in different session

Test Plan: buck test //caffe2/torch/csrc/deploy:test_deploy

Reviewed By: suo

Differential Revision: D31323045

fbshipit-source-id: 2f55a6a042130afa29e73da768632ed78b90e092
